### PR TITLE
refactor: enable -Wconversion warning for syntax.c

### PR DESCRIPTION
**Pull request by dundargoc**
Work on https://github.com/neovim/neovim/issues/567

### CODE REVIEW

File: src/nvim/syntax.c [hunk 1]
```diff
@@ -430,7 +430,7 @@ void syntax_start(win_T *wp, linenr_T lnum)
     } else {
       // Need to parse "minlines" lines before state can be considered
       // valid to store.
-      first_stored = current_lnum + syn_block->b_syn_sync_minlines;
+      first_stored = current_lnum + (linenr_T)syn_block->b_syn_sync_minlines;
```
Comment (zeertzjq) [inline src/nvim/syntax.c:433]: Will it make more sense to check the three `long`s in `synblock_T` to `int32_t`?

Comment (dundargoc) [reply src/nvim/syntax.c]: Sounds good.

File: src/nvim/syntax.c [hunk 2]
```diff
@@ -994,11 +994,10 @@ void syn_stack_free_all(synblock_T *block)
  */
 static void syn_stack_alloc(void)
 {
-  long len;
   synstate_T *to, *from;
   synstate_T *sstp;
 
-  len = syn_buf->b_ml.ml_line_count / SST_DIST + Rows * 2;
+  int len = syn_buf->b_ml.ml_line_count / SST_DIST + Rows * 2;
```
Comment (zeertzjq) [inline src/nvim/syntax.c:1000]: Is it safe to assume `int` is at least 32 bits?

Comment (dundargoc) [reply src/nvim/syntax.c:1000]: yes

File: src/nvim/syntax.c [hunk 3]
```diff
@@ -5211,11 +5203,11 @@ static void syn_cmd_sync(exarg_T *eap, int syncing)
       n = getdigits_long(&arg_end, false, 0);
```
Comment (zeertzjq) [inline src/nvim/syntax.c:5203]: This `n` can probably be changed to `int32_t` or `linenr_T` and use `getdigits_int32` here.

File: src/nvim/syntax.c [hunk 4]
```diff
@@ -5879,7 +5871,7 @@ int syn_get_foldlevel(win_T *wp, long lnum)
   if (wp->w_s->b_syn_folditems != 0
       && !wp->w_s->b_syn_error
       && !wp->w_s->b_syn_slow) {
-    syntax_start(wp, lnum);
+    syntax_start(wp, (linenr_T)lnum);
```
Comment (zeertzjq) [inline src/nvim/syntax.c:5874]: This `lnum` argument is only passed a `linenr_T`, so it can be changed to a `linenr_T`.

File: src/nvim/syntax.c [hunk 5]
```diff
@@ -5791,7 +5783,7 @@ int syn_get_id(win_T *wp, long lnum, colnr_T col, int trans, bool *spellp, int k
   // When the position is not after the current position and in the same
   // line of the same window with the same buffer, need to restart parsing.
   if (wp != syn_win || wp->w_buffer != syn_buf || lnum != current_lnum || col < current_col) {
-    syntax_start(wp, lnum);
+    syntax_start(wp, (linenr_T)lnum);
```
Comment (zeertzjq) [inline src/nvim/syntax.c:5786]: This `lnum` argument is only passed a `linenr_T`, so it can be changed to a `linenr_T`.

File: src/nvim/syntax.c [hunk 6]
```diff
@@ -1084,7 +1083,7 @@ static void syn_stack_apply_changes_block(synblock_T *block, buf_T *buf)
   prev = NULL;
   for (p = block->b_sst_first; p != NULL;) {
     if (p->sst_lnum + block->b_syn_sync_linebreaks > buf->b_mod_top) {
-      n = p->sst_lnum + buf->b_mod_xlines;
+      n = p->sst_lnum + (linenr_T)buf->b_mod_xlines;
```
Comment (zeertzjq) [inline src/nvim/syntax.c:1086]: `b_mod_xlines` can be changed to `int32_t` or `linenr_T`.

File: src/nvim/syntax.c [hunk 7]
```diff
@@ -2916,9 +2914,8 @@ static int syn_regexec(regmmatch_T *rmp, linenr_T lnum, colnr_T col, syn_time_T
     return false;
   }
 
-  rmp->rmm_maxcol = syn_buf->b_p_smc;
-  r = vim_regexec_multi(rmp, syn_win, syn_buf, lnum, col,
-                        syn_tm, &timed_out);
+  rmp->rmm_maxcol = (colnr_T)syn_buf->b_p_smc;
+  int r = (int)vim_regexec_multi(rmp, syn_win, syn_buf, lnum, col, syn_tm, &timed_out);
```
Comment (zeertzjq) [inline src/nvim/syntax.c:2918]:
```suggestion
  long r = vim_regexec_multi(rmp, syn_win, syn_buf, lnum, col, syn_tm, &timed_out);
```

**Pull request merged**